### ISSUE TITLE
Parallelise OCDS  Work

### ISSUE THREAD

**Issue by odscjames** [open-contracting/kingfisher#197]
(I thought there was another ticket about this, sorry if dupe)

There have been various discussions among people, including at the Montreal retreat about this.

We would like to parallelise the work. There are several different places this could happen

A) Within stages. eg, currently only one fetch happens at a time, only one store happens at a time
It might be nice to parallelise so up to a certain number (eg 5) could run at once. However, this basically means we are DOS'ing someones server. During the fetch stage, it's someone elses. During  store stage, it's our own. So careful. (Set limits per source?)

B) Across stages. Currently you can't do any storing at all until all the fetch has finished.
Some sources are very big (5000 files, more) and it would be good if as soon as 1 file is fetched, it it stored. And as soon as 1 file is stored, it's it Checked. This means analysts can start work sooner.

Also, if we get this right it will hopefully make #58 really easy.

Tech to use:
Options: 

- Threads 
- Async 
- A que system and lots of separate workers.

The consensus seems to the latter: A que system and tons of separate workers.

- While Async might give us even more of a speed boost for better use of our server resources, it's a lot of extra work.
- Async and threads are seen to be more work than a que system
- It can be hard to debug problems with Async and threads, where as if we get a que system right (transparent) it should be easier.
- It might be easier to make a que system resumable  
- A que system would support spreading workers across multiple machines easily

It was not discussed which que system to use, some of the team have "opinions" :-)

Also noted is that ideally it should be able to work without the que in an "single threaded" way, so as not to increase the install burden.

This is a big bit of work and it will **not** start straight away; this is simply to record conversation and thoughts so far and allow others to comment on this.

**Comment by robredpath** [comment]
In the very short term, it would be useful to have explicitly recorded what can and can't be done in parallel. As I understand it, I could open up 20 tmux panes and run all stages for each source individually with no ill effects (apart from making the server very busy). If I was trying to get hold of all the data as quickly as possible, this would help me a lot - although wouldn't help in the case of trying to get one source's data more quickly. 

In the merely short term, I'd like us to consider the trade-off between 'doing it right' and 'making it immediately useful'. Right now, running some of the scrapers takes a very long time, which is an immediate pain point. If the longest-running ones can be made to download in parallel using an easier-to-try technique, then we solve an immediate problem sooner rather than later. We can, of course, then refactor to a better technique later.

**Comment by odscjames** [comment]
> In the very short term, it would be useful to have explicitly recorded what can

Sounds like a good docs page to write up - thanks for prompt!

**Comment by jpmckinney** [comment]
There was a very brief exchange about using state machines in https://github.com/open-contracting/kingfisher/commit/f776d7a38a1351d41f00e53246fc8c6a20092807#r30481553

Note that many queues are essentially (very simple) state machines, so this is essentially on the broad path of the issue description.

In terms of the database, to implement a state machine, we just need a `state` column in any table for which there is a state machine.

In terms of code, here's an [example state machine API](https://github.com/pluginaweek/state_machine#usage). If this model, we could have different workers trigger `gather` events on sources, and `fetch` and `store` events on files (among other events). In a state machine, each event describes its valid transitions, e.g. `store` transitions from `fetched` to `stored`, and triggering `store` on a file in a state other than `fetched` would raise an error.

You could of course implement all this without the language of a 'state machine', but in my experience, using this model keeps logic, responsibility, concerns, etc. tidy.

At present, I think we only run 'events' at the level of the source (and we only have a `Source` class, not a `File` or `Resource` class), but I think we have to move this down to the level of the file to fulfill the vision of parallelization, which also gives an opportunity to implement these state machines.

Somewhat separately, I find 'filestatus' a confusing name, as the table describes a file/resource, not just a file status.

**Comment by odscjames** [comment]
This is handled for us by a combination of
* the 2 parts in the new system allowing data to be stored and (in some cases) processed as soon as it is fetched
* scrapy does parallelisation for us
closing
